category_theory/functor.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2017 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tim Baumann, Stephen Morgan, Scott Morrison
+
+Defines a functor between categories.
+
+(As it is a 'bundled' object rather than the `is_functorial` typeclass parametrised 
+by the underlying function on objects, the name is capitalised.)
+
+Introduces notations
+  `C ↝ D` for the type of all functors from `C` to `D`. (I would like a better arrow here, unfortunately ⇒ (`\functor`) is taken by core.)
+  `F X` (a coercion) for a functor `F` acting on an object `X`.
+-/
+
+import .category
+
+namespace category_theory
+ 
+universes u₁ v₁ u₂ v₂ u₃ v₃
+
+structure functor (C : Type u₁) [category.{u₁ v₁} C] (D : Type u₂) [category.{u₂ v₂} D] : Type (max u₁ v₁ u₂ v₂) :=
+(obj           : C → D)
+(map           : Π {X Y : C}, (X ⟶ Y) → ((obj X) ⟶ (obj Y)))
+(map_id        : ∀ (X : C), map (𝟙 X) = 𝟙 (obj X) . obviously)
+(functoriality : ∀ {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z), map (f ≫ g) = (map f) ≫ (map g) . obviously)
```

Comment:
I think I will insist on this one. I know that literal description of axioms doesn't appeal to you, but there are just too many different ways of phrasing lemmas and it is a *huge* benefit to users that are bouncing between fields of mathematics that they may not know in detail the "traditional" way, to be able to guess names based solely on how they are written rather than what they mean. As a simple example, the symmetric version of this equality would also be called "functoriality" by a mathematician, as well as a hypothetical definition like `commutes map (≫)` which unfolds to this equality. You have `map` and `comp` being applied; so it is called `map_comp`.